### DISCUSSION TITLE
`__rsub__` behaviour when using bounded `TypeVar`.

### DISCUSSION THREAD

**Discussion by MaicoTimmerman** [microsoft/pyright#7321] [Q&A]
The typeshed definition for `dateutil.relativedelta.__rsub__` is defined as:

```python
_DateT = TypeVar("_DateT", date, datetime)

# I presumed the above definition was wrong, but setting an upper bound on the typevar also doesn't work.
# _DateT = TypeVar("_DateT", bound=date)  # datetime subclasses date


class relativedelta:
    # omitted for brevity
    @overload
    def __rsub__(self, other: A) -> Self: ...
    @overload
    def __rsub__(self, other: timedelta) -> Self: ...
    @overload
    def __rsub__(self, other: _DateT) -> _DateT: ...
```

Now I was working with a subclass of `datetime`, imagine it looks like:

```
class MyDateTime(datetime):
    pass
```
Now, is there a way to get pyright to return `MyDateTime` for the following expressions?

```python
d = MyDateTime()
x = d - relativedelta(days=1)
assert_type(x, MyDateTime)
```

**Comment by erictraut** [comment]
Could you please provide a minimal, self-contained code sample? Your code snippets above refer to a number of undefined symbols, so it's difficult to reconstruct your repro.

**Reply by MaicoTimmerman** [reply]
My question was based on the failures I was observing in [this typeshed MR](https://github.com/python/typeshed/pull/11463/files):

```python
from datetime import date, datetime
from typing_extensions import assert_type

from dateutil.relativedelta import relativedelta


class MyDateTime(datetime):
    pass


d = MyDateTime.now()
x = d - relativedelta(days=1)
assert_type(x, MyDateTime)
```

I'll have to dive somewhat deeper, as I'm not able to reproduce the CI failure locally using pyright==1.1.342.
